File: src/main/scala/io/arabesque/report/MasterReport.scala
package io.arabesque.report

import scala.collection.mutable.{ArrayBuffer, StringBuilder}
import java.io._

import scala.collection.mutable

/**
  * Created by ehussein on 7/31/17.
  */
class MasterReport extends EngineReport {
  var storageSummary: ArrayBuffer[String] = new ArrayBuffer[String]()
  var storageSize: ArrayBuffer[Long] = new ArrayBuffer[Long]()
  var patternSize: ArrayBuffer[Long] = new ArrayBuffer[Long]()

  override def saveReport(path: String) = {
    val filePath = s"$path/MasterReport.txt"

    if(superstep == 0) {
      val file = new File(filePath)
      if(file.exists())
        file.delete()
    }
    else {
      // to remove "]" written at last superstep
      val raf: RandomAccessFile = new RandomAccessFile(new File(filePath), "rw")
      raf.setLength(raf.length() - 1)
      raf.close()
    }

    val pw: PrintWriter = new PrintWriter(new BufferedWriter(new FileWriter(filePath,true)))

    if(superstep == 0) {
      pw.print("[")
    }
    else
      pw.print(",")

    val report = toString()

    pw.print(report + "\n")
    pw.print("]")

    pw.close()
  }

  override def toString(): String = toJSONString()

  def toJSONString(): String = {
    val str: StringBuilder = new mutable.StringBuilder()

    str.append(s"""{\"super_step\":$superstep, """)
    str.append(s"""\"runtime\":$getRuntime, """)
    str.append("\"StorageSummary\":[")

    var i = 0
    while(i < storageSize.length) {
      str.append(s"""{\"Summary\":${storageSummary(i)},""")
      str.append(s"""\"StorageSize\":${storageSize(i)},""")
      str.append(s"""\"PatternSize\":${patternSize(i)}}""")

      i += 1

      if(i != storageSize.length)
        str.append(", ")
    }

    str.append("]")
    str.append("}")

    str.toString()
  }
}

File: src/main/java/io/arabesque/report/MasterReport.java
package io.arabesque.report;

import java.io.IOException;
import java.util.ArrayList;
/**
 * Created by ehussein on 10/1/17.
 */
public class MasterReport extends EngineReport {
    public ArrayList<String> storageSummary = new ArrayList<String>();
    public ArrayList<Long> patternSize = new ArrayList<Long>();
    public ArrayList<Long> storageSize = new ArrayList<Long>();
    public ArrayList<Long> domainEntriesCalculatedSize = new ArrayList<Long>();
    public ArrayList<Long> calculatedSize = new ArrayList<Long>();
    public ArrayList<Long> numberOfWordsInDomains = new ArrayList<Long>();
    public ArrayList<Long> numberOfWordsInConnections = new ArrayList<Long>();
    protected long superstepStorageSize = 0;
    protected long superstepPatternSize = 0;
    protected long superstepCalculatedSize = 0;
    protected long superstepDomainEntriesCalculatedSize = 0;

    protected void calcCummulatives() {
        superstepStorageSize = 0;
        superstepPatternSize = 0;
        superstepCalculatedSize = 0;
        superstepDomainEntriesCalculatedSize = 0;

        for(int i = 0 ; i < storageSummary.size() ; ++i) {
            superstepStorageSize += storageSize.get(i);
            superstepPatternSize += patternSize.get(i);
            superstepCalculatedSize += calculatedSize.get(i);
            superstepDomainEntriesCalculatedSize += domainEntriesCalculatedSize.get(i);
        }
    }

    @Override
    public void saveReport(String path) throws IOException {
        ensurePathExists(path);
        String filePath = path + "/MasterReport.txt";
        super.saveReport(filePath);
    }

    @Override
    public String toString() { return toJSONString(); }

    public String toJSONString() {
        calcCummulatives();

        StringBuilder str = new StringBuilder();

        str.append("{\"super_step\":" + superstep + ", ");
        str.append("\"runtime\":" + getRuntime() + ", ");
        str.append("\"TotalStorageSize\":" + superstepStorageSize + ", ");
        str.append("\"TotalPatternSize\":" + superstepPatternSize + ", ");
        str.append("\"TotalCalculatedSize\":" + superstepCalculatedSize + ", ");
        str.append("\"TotalDomainEntriesCalculatedSize\":" + superstepDomainEntriesCalculatedSize + ", ");
        str.append("\"StorageSummary\":[");

        int i = 0;
        while(i < storageSize.size()) {
            str.append("{");
            str.append("\"Summary\":" + storageSummary.get(i) + ",");
            str.append("\"NumberOfWordsInDomains\":" + numberOfWordsInDomains.get(i) + ",");
            str.append("\"NumberOfWordsInConnections\":" + numberOfWordsInConnections.get(i) + ",");
            str.append("\"PatternSize\":" + patternSize.get(i) + ",");
            str.append("\"StorageSize\":" + storageSize.get(i) + ",");
            str.append("\"DomainEntriesCalculatedSize\":" + domainEntriesCalculatedSize.get(i) + ",");
            str.append("\"CalculatedSize\":" + calculatedSize.get(i));
            str.append("}");

            i += 1;

            if(i != storageSize.size())
                str.append(", ");
        }

        str.append("]");
        str.append("}");

        return str.toString();
    }
}
